ddht/v5_1/alexandria/client.py

```diff
@@ -0,0 +1,130 @@
+from typing import Any, Optional, Set, Type
+
+from eth_typing import NodeID
+
+from ddht.endpoint import Endpoint
+from ddht.exceptions import DecodingError
+from ddht.request_tracker import RequestTracker
+from ddht.v5_1.abc import NetworkAPI
+from ddht.v5_1.alexandria.abc import AlexandriaClientAPI
+from ddht.v5_1.alexandria.constants import ALEXANDRIA_PROTOCOL_ID
+from ddht.v5_1.alexandria.messages import (
+    AlexandriaMessage,
+    PingMessage,
+    PongMessage,
+    TAlexandriaMessage,
+    decode_message,
+)
+from ddht.v5_1.alexandria.payloads import PingPayload, PongPayload
+
+
+class AlexandriaClient(AlexandriaClientAPI):
+    protocol_id = ALEXANDRIA_PROTOCOL_ID
+
+    _active_request_ids: Set[bytes]
+
+    def __init__(self, network: NetworkAPI) -> None:
+        self.network = network
+        self.request_tracker = RequestTracker()
+
+        self._active_request_ids = set()
+
+    #
+    # Request/Response message sending primatives
+    #
+    async def _send_request(
+        self,
+        node_id: NodeID,
+        endpoint: Endpoint,
+        message: AlexandriaMessage[Any],
+        *,
+        request_id: Optional[bytes] = None,
+    ) -> bytes:
+        data_payload = message.to_wire_bytes()
+        request_id = await self.network.client.send_talk_request(
```

Comment:
*nit*: This looks a little strange, pulling out a member of a member breaks the law of demeter and makes this code a little more fragile during refactorings. Something here isn't quite factored right though I'm not sure what.

Looking through the rest of this class, it seems like it almost exclusively uses the `Client`. It calls `self.network.talk` but that's just a passthrough to the client. Maybe this class could be passed the client and the enr_manager?

Reply:
First, :heart: learning new terminology.  I wasn't aware of https://en.wikipedia.org/wiki/Law_of_Demeter

Depending on how strongly you feel about this. I'm -0 on fixing this mostly because I'm not sure how to without a lot of futzing against an API that is currently functioning very well.  The `ClientAPI` gives low level un-opinionated APIs for accessing the *raw* protocol.  The `NetworkAPI` exposes the business logic and hides away the internals and only exposes high level stuff.  What I found while building all of this out was that there were times when you needed each of these and that this separation makes testing easier and provides good encapsulation of these two tiers of functionality.

So let me know if either you feel strongly about this..